### PR TITLE
Feature/part resize: Add key bindings to resize focused view

### PR DESCRIPTION
**Pull request by cleidigh**
See issue #22645 

Part Resize Feature Notes:

Proposed feature to address issue #22645:
	Keyboard bindings to expand or contract the main axis of the focused view

Two key bindings expand view, contract view

key bindings based on part focus (potential reduction of key collisions)

expansion or contraction follow main axis of part and orientation

eg sidebar has focus, expand moves border right if on left
or left if sidebar on right
eg panel has focus expand moves border up...
editor focus: expand moves right border right (h. layout)
top up (v. layout) or opposite if single editor
approach still works if a new part eg secondary sidebar
is added in the future

Possible (mnemonic) key bindings: (probably none assigned by default)

[expand part]: ctrl+alt+shift+ >
[contract part]: ctrl+alt+shift+ <

[expand part]: ctrl+alt+shift+ x
[contract part]: ctrl+alt+shift+ c

[expand part]: ctrl+alt+shift+ )
[contract part]: ctrl+alt+shift+ (

[restore default part size]: ctrl+alt+shift+ d
[restore default part size]: ctrl+alt+shift+ r

Potential Future Issues:

A part with no "primary axis" (expanding contract ambiguous)

Main file work:

Main file, follows toggleSideBarVisibility.ts design approach
- workbench command registrations
- action functions

.\src\vs\workbench\browser\actions\resizePart.ts

Workbench Support Functions
- set up per Part
- access to workbench layout
.\src\vs\workbench\electron-browser\workbench.ts

Layout Support Functions
- adjust main access dimensions pre- layout call
- possible reset to default option
.\src\vs\workbench\browser\layout.ts

The prototype implementation sports resize
of the sidebar, the panel and one editor.

After looking more into the code base as well
as the navigation contribution by @misoguy
I think some files need to be moved around.
my first implementation was based on toggleSideBarVisibility
but I think the more recent code  add
the commands to actions.ts

The next step is to flush out support for the editors
where more than one is visible. and re- factor
per suggestion input.

Thanks for considering...

### CODE REVIEW

**Comment by cleidigh** [comment]
my apologies and get hub ignorance
I have to exclude the several local working files
like tasks.json etc

**Comment by cleidigh** [comment]
okay I think I have this correct now

File: src/vs/workbench/browser/actions/resizePart.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		}
+		else if (isEditorFocus) {
+			// console.log('editor f1ocus');
+			return this.partService.resizePart(Parts.EDITOR_PART, RESIZE_INCREMENT);
+		}
+		return TPromise.as(false);
+	}
+}
+
+export class ContractViewAction extends Action {
```
Comment (misoguy) [inline src/vs/workbench/browser/actions/resizePart.ts]: It seems to me that a lot of the code in ```ContractViewAction``` looks the same with ```ExpandViewAction``` except for the increment value which could be abstracted to a ```BaseResizeViewAction``` and an argument ```INCREMENT_DIRECTION``` (probably 1 or -1) could be passed down.

Comment (cleidigh) [inline src/vs/workbench/browser/actions/resizePart.ts]: @misoguy 
Agree, I've looked at your code a lot as well as comments. I was anxious
to get some initial feedback before doing more re- factoring.
I also think moving the resize actions into actions.ts like your code
is the newer approach. When I started I modeled my code after the toggleSidebarVisibility.ts
not sure why these several toggle functions are in separate files, anybody?

Hope you give it a try, seems like we have some common interests...

**Comment by cleidigh** [comment]
Forgot Note: 

Temporary Keybindings for testing:

expandView: ctrl+alt+b
contractView: ctrl+alt+e

>

**Comment by bpasero** [comment]
I am currently on business travel and will get back to your PR once I find some time, sorry for not being more responsive at the moment 👍

**Comment by cleidigh** [comment]
@bpasero 
no problemo...

I'm diving into the group and editor classes to determine the best way to manipulate them.

File: src/vs/workbench/browser/actions/resizePart.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import { IWorkbenchActionRegistry, Extensions } from 'vs/workbench/common/actionRegistry';
+import { IPartService, Parts } from 'vs/workbench/services/part/common/partService';
+import { KeyMod, KeyCode } from 'vs/base/common/keyCodes';
+
+const RESIZE_INCREMENT = 50;
+
+export class ExpandViewAction extends Action {
+
+	public static ID = 'workbench.action.expandView';
+	public static LABEL = nls.localize('expandView', "Expand Current View");
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/resizePart.ts]: I suggest to name this differently, e.g. "Increase Current View Size" and `workbench.action.increaseViewSize`

File: src/vs/workbench/browser/actions/resizePart.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+	) {
+		super(id, label);
+
+		this.enabled = !!this.partService;
+	}
+
+	public run(): TPromise<any> {
+
+		const isEditorFocus = this.partService.hasFocus(Parts.EDITOR_PART);
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/resizePart.ts]: This logic should be extracted into a base class and shared across both actions.

File: src/vs/workbench/browser/actions/resizePart.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+			return this.partService.resizePart(Parts.EDITOR_PART, -RESIZE_INCREMENT);
+		}
+		return TPromise.as(false);
+	}
+}
+
+// probably won't have default key bindings, set up for easy test...
+
+const registry = Registry.as<IWorkbenchActionRegistry>(Extensions.WorkbenchActions);
+registry.registerWorkbenchAction(new SyncActionDescriptor(ExpandViewAction, ExpandViewAction.ID, ExpandViewAction.LABEL, { primary: KeyMod.CtrlCmd | KeyMod.Alt | KeyCode.KEY_B }), 'View: Expand View', nls.localize('view', "View"));
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/resizePart.ts]: Are these keybindings just for testing?

File: src/vs/workbench/browser/layout.ts [hunk 5]
```diff
@@ -539,6 +539,48 @@ export class WorkbenchLayout implements IVerticalSashLayoutProvider, IHorizontal
 		return this.panelMaximized;
 	}
 
+	// Add reset for keyboard support
+	public resetSideBarWidth(): number {
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: Not being used?

File: src/vs/workbench/browser/layout.ts [hunk 6]
```diff
@@ -539,6 +539,48 @@ export class WorkbenchLayout implements IVerticalSashLayoutProvider, IHorizontal
...
+		this.sidebarWidth = MIN_SIDEBAR_PART_WIDTH;
+		return this.sidebarWidth;
+	}
+
+	// change part size along the main axis
+	public setPartSizeChange(part: Parts, sizeChange: number): boolean {
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: This method should do the same validation about minimal and maximal sizes for a specific part as we already do when the user resizes a view using the sashes (e.g. [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/browser/layout.ts#L192))

File: src/vs/workbench/electron-browser/workbench.main.ts [hunk 7]
```diff
@@ -25,6 +25,7 @@ import 'vs/workbench/browser/actions/toggleSidebarVisibility';
 import 'vs/workbench/browser/actions/toggleSidebarPosition';
 import 'vs/workbench/browser/actions/toggleEditorLayout';
 import 'vs/workbench/browser/actions/toggleZenMode';
+import 'vs/workbench/browser/actions/reSizePart';
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.main.ts]: Lowercase 's' to be `resizePart` ?

File: src/vs/workbench/electron-browser/workbench.ts [hunk 8]
```diff
@@ -1142,6 +1142,24 @@ export class Workbench implements IPartService {
 		}
 	}
 
+	// Resize requested part along the main axis
+	// layout will do all the math for us and adjusts the other Parts
+	public resizePart(part: Parts, sizeChangePx: number, skipLayout?: boolean): TPromise<void> {
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: This method does not need to return a promise and also does not seem to be using `skipLayout`. I would suggest to do the `layout()` call from within `layout.ts` and not outside.

**Review by bpasero — CHANGES_REQUESTED** [review]
Some initial feedback provided, good start 👍

**Comment by cleidigh** [comment]
@bpasero 

Thanks for taking the time to review the code and comment (sorry for some of the newbie mistakes)

all your comments make sense I will check them off and
will comment on your points in line and add my questions / issues here - let me know 
if it's better just to do everything in one comment in the future...

I have a fully working prototype I'm pretty happy with, but I do have some issues
questions related to structure and code location.  

Overview of current implementation:
- can expand and contract focused sidebar or panel
- can expand and contract one two or three visible editors
- if only one editor , expand reduces sidebar if visible
- for two or three editors expansion increases focused editors
- the remaining editors expand or contract according to their current percentages
- I added one function to editorGroupsControl that has visibility to the editors
 
 My issues and questions:
 - looking at other code, I think I should move code from resizePart.ts into actions.ts
 	agree ?

- I would like all resize functions and layout calls to occur in layout.ts
  To accomplish that I have to import Editor into layout.ts to access
  editorGroupsControl, unless I can expose it in another fashion.
    agreed ?

- Currently I let the layout functions do the final adjustments and size checks
  (goes to one of your comments) I wasn't sure about duplicate checks.
  one issue I have, or behavior I don't like is that if one expands
  the sidebar and any editor hits its minimum, the expansion of the sidebar stops
  I can fix this by stealing from an editor that is still larger than its minimum.
  I will just have to add logic to check and fix.
  I think this is the right to change to do to my current implementation
    agree ?

- Currently I use a fixed pixel adjustment 
	I think I should change to a percentage screen media size - agree ?
	if we keep this fixed for now (not user adjustable)	what percentage ? 7.5% 
	is my thought - yours ?
	I'm okay with adding the user setting if desired - I just want to hits
	the April release

I'm hoping to have all your corrections and my new code by tomorrow
assuming I know any structural changes.

thanks again for your help !

File: src/vs/workbench/browser/actions/resizePart.ts (hunk 2 above)
Comment (cleidigh) [inline src/vs/workbench/browser/actions/resizePart.ts]: sounds fine

File: src/vs/workbench/browser/actions/resizePart.ts (hunk 3 above)
Comment (cleidigh) [inline src/vs/workbench/browser/actions/resizePart.ts]: already in progress

File: src/vs/workbench/browser/actions/resizePart.ts (hunk 4 above)
Comment (cleidigh) [inline src/vs/workbench/browser/actions/resizePart.ts]: just for testing
final code will leave commands unbound

File: src/vs/workbench/browser/layout.ts (hunk 5 above)
Comment (cleidigh) [inline src/vs/workbench/browser/layout.ts]: I originally thought of a third command to reset to defaults
I don't think that's necessary, need to remove

File: src/vs/workbench/browser/layout.ts (hunk 6 above)
Comment (cleidigh) [inline src/vs/workbench/browser/layout.ts]: should not be a problem - relates to one of my behavior issues below
I want to fix also

File: src/vs/workbench/electron-browser/workbench.main.ts (hunk 7 above)
Comment (cleidigh) [inline src/vs/workbench/electron-browser/workbench.main.ts]: dictation error ;-)
not sure why this passed compile and lint

File: src/vs/workbench/electron-browser/workbench.ts (hunk 8 above)
Comment (cleidigh) [inline src/vs/workbench/electron-browser/workbench.ts]: I have to delve into the promises use better
I guess it is just the layout function that needs the promise structure

File: src/vs/workbench/browser/actions/resizePart.ts (hunk 2 above)
Comment (cleidigh) [inline src/vs/workbench/browser/actions/resizePart.ts]: done

File: src/vs/workbench/browser/actions/resizePart.ts (hunk 3 above)
Comment (cleidigh) [inline src/vs/workbench/browser/actions/resizePart.ts]: done

still have to decide if this should go into actions.ts

File: src/vs/workbench/browser/layout.ts (hunk 5 above)
Comment (cleidigh) [inline src/vs/workbench/browser/layout.ts]: killed

File: src/vs/workbench/browser/layout.ts (hunk 6 above)
Comment (cleidigh) [inline src/vs/workbench/browser/layout.ts]: have most of this
still I have to tweak for editors

File: src/vs/workbench/electron-browser/workbench.main.ts (hunk 7 above)
Comment (cleidigh) [inline src/vs/workbench/electron-browser/workbench.main.ts]: done

File: src/vs/workbench/electron-browser/workbench.ts (hunk 8 above)
Comment (cleidigh) [inline src/vs/workbench/electron-browser/workbench.ts]: change to void
layout in layout

**Comment by cleidigh** [comment]
okay I think I know how to do the class structure
I need to add the editor\group resize into editorGroupService
that way I can do everything in layout.ts
much cleaner, yes?

should be able to push tomorrow

**Comment by bpasero** [comment]
> looking at other code, I think I should move code from resizePart.ts into actions.ts
> agree ?

Yes, fine by me.

> I would like all resize functions and layout calls to occur in layout.ts
> To accomplish that I have to import Editor into layout.ts to access
> editorGroupsControl, unless I can expose it in another fashion.
> agreed ?

I suggest to find out from the action if the resizing happens within the editor group or not and then delegate the call to the IEditorGroupService and do it within. I think this is going in a hierarchy:
* if user is outside editor, the resizing can happen in layout.ts
* if user is inside editor, the resizing should happen within editorGroupsControl that is responsible for laying out editors (layout.ts does not, only the outer container)

> Currently I use a fixed pixel adjustment
> I think I should change to a percentage screen media size - agree ?
> if we keep this fixed for now (not user adjustable)	what percentage ? 7.5%
> is my thought - yours ?
> I'm okay with adding the user setting if desired - I would hope to hit
> the April release

I think a final pixel value is OK to start with.

**Comment by cleidigh** [comment]
- did the percentage change yesterday (basically one line/you decide)
- I'm doing just what you suggested already with delegating sidebar and panel to lay out
and editor resizing to the editorGroupsControl - just have to get minimax checks in
thanks

**Comment by cleidigh** [comment]
@bpasero 
quick update
silo size checking took a little longer than expected
everything else just checked off just have to move to actions
then I'll push my commits, hopefully in a couple hours

**Comment by cleidigh** [comment]
@bpasero 
okay this is ready for review
I'm sure there will be things to address
but I want to see I have nothing structurally wrong
and you have a chance to see the behaviors

note : I made the choice to favor the focused part and increase  the editor size, shrinking both
the others if there are three editors. Maintaining starting ratio. This happens until minimum
sizes met. If you contract after this the ratios may have been lost an equal distribution
will occur.

File: src/vs/workbench/browser/parts/editor/editorGroupsControl.ts [hunk 9]
```diff
@@ -78,7 +78,11 @@ export interface IEditorGroupsControl {
 	setGroupOrientation(orientation: GroupOrientation): void;
 	getGroupOrientation(): GroupOrientation;
 
+	requestActiveGroupSizeChange(groupSizeChange: number): boolean;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorGroupsControl.ts]: Can we maybe call this similar to the other one `resizePart` in `partService`, e.g. `resizeGroup` and you pass in the position of the group to resize?

File: src/vs/workbench/electron-browser/workbench.ts [hunk 10]
```diff
@@ -1142,6 +1142,24 @@ export class Workbench implements IPartService {
...
 
+	// Resize requested part along the main axis
+	// layout will do all the math for us and adjusts the other Parts
+	public resizePart(part: Parts, sizeChange: number): void {
+
+		switch (part) {
+			case Parts.SIDEBAR_PART:
+			case Parts.PANEL_PART:
+			case Parts.EDITOR_PART:
+				this.workbenchLayout.setPartSizeChange(part, sizeChange);
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: `setPartSizeChange` => `resizePart` ?

File: src/vs/workbench/browser/parts/editor/editorGroupsControl.ts [hunk 11]
```diff
@@ -853,6 +857,83 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
...
+				}
+			default:
+				break;
+		}
+		this.layout(this.dimension);
+
+		return true;
+	}
+
+	protected boundSiloSize(siloPosition: Position, sizeChangePx: number): number {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorGroupsControl.ts]: `protected` =>  `private`?

File: src/vs/workbench/browser/parts/editor/editorGroupsControl.ts [hunk 12]
```diff
@@ -853,6 +857,83 @@ export class EditorGroupsControl extends Themable implements IEditorGroupsContro
...
+		const visibleEditors = this.getVisibleEditorCount();
+		let newSiloSize: number = 0;
+
+		newSiloSize = Math.max(this.minSize, this.silosSize[siloPosition] + sizeChangePx);
+		newSiloSize = Math.min(newSiloSize, (this.totalSize - this.minSize * (visibleEditors - 1)));
+
+		return newSiloSize;
+	}
+
+	protected distributeRemainingSilosSize(remPosition1: Position, remPosition2: Position, availableSize: number): void {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorGroupsControl.ts]: `protected` =>  `private`?

**Review by bpasero — CHANGES_REQUESTED** [review]
@cleidigh good job! this seems to work very nice, a couple of feedback attached 👍

File: src/vs/workbench/browser/parts/editor/editorGroupsControl.ts (hunk 9 above)
Comment (cleidigh) [inline src/vs/workbench/browser/parts/editor/editorGroupsControl.ts]: done
was never wedded to the name ;-)

File: src/vs/workbench/browser/parts/editor/editorGroupsControl.ts (hunk 11 above)
Comment (cleidigh) [inline src/vs/workbench/browser/parts/editor/editorGroupsControl.ts]: done

File: src/vs/workbench/browser/parts/editor/editorGroupsControl.ts (hunk 12 above)
Comment (cleidigh) [inline src/vs/workbench/browser/parts/editor/editorGroupsControl.ts]: done

File: src/vs/workbench/electron-browser/workbench.ts (hunk 10 above)
Comment (cleidigh) [inline src/vs/workbench/electron-browser/workbench.ts]: done

**Comment by cleidigh** [comment]
F2 helps a lot !
but it seemed to miss some files, maybe there's a scope thing I don't understand

**Comment by bpasero** [comment]
@cleidigh excellent job 👍

**Comment by cleidigh** [comment]
@bpasero 
awesome thanks - just happened to see you added the debug support for the shared process - will try out today

fyi - working on this project is very good for me. since I'm not able to work a regular job I'm very interested in becoming regular contributor. I'm not fast but I do have a decent amount of time
to devote to this. given my firmware/communication software experience I'm hoping to be able to
help with some of the lower level process/debugging/issues eventually. I also have found a few bugs in SCM I'm looking at now.

In the meantime I'm more than happy to help out with issues/bugs either for the March deadline
for upcoming April deadline . I know and have already looked at the help-wanted issues, however
I don't know what the priorities are. if you would like to give me some assignments that you think
would be useful and within my range , I'm all up for it. and thanks for working to review and to merge
the new code !

This participation and focused goals are a great thing for me...

cheers Christopher

**Comment by bpasero** [comment]
@cleidigh yes, I added a debug configuration that allows to debug both main and renderer process from within VS Code (provided you installed the Chrome Debug Adapter extension).

Thanks for willing to help out more, going after the [help wanted](https://github.com/Microsoft/vscode/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22+sort%3Areactions-%2B1-desc) bugs is a good start. Often it helps to first quickly ask if it is OK to start working on a particular issue, just to confirm that we have cycles to review a potential contribution in that area and it makes sense via PR.

As for priorities, we typically just go by number of votes on an issue to decide if something is important or not :)

**Comment by cleidigh** [comment]
@bpasero 

- tested newly built OSS
- see my code - cool
- testing debugging - appears  that runtimeArg $(workspaceRoot) presumably replaced
by webroot debugging parameter prevents launch (alone or in compound)
- remove and launch\attach and debugging work

{
			"type": "chrome",
			"request": "launch",
			"name": "Launch VS Code",
			"windows": {
				"runtimeExecutable": "${workspaceRoot}/scripts/code.bat"
			},
			"osx": {
				"runtimeExecutable": "${workspaceRoot}/scripts/code.sh"
			},
			"linux": {
				"runtimeExecutable": "${workspaceRoot}/scripts/code.sh"
			},
			"urlFilter": "*index.html*",
			"runtimeArgs": [
***				"${workspaceRoot}",
				"--debug=5875"
			],
			"webRoot": "${workspaceRoot}"
		}

**Comment by bpasero** [comment]
Weird, it really works for me out of the box. Are you using our insiders build or stable?

**Comment by cleidigh** [comment]
@bpasero 
was using last nights Insiders
just tried stable - same problem - have to remove run time arg to work

**Comment by bpasero** [comment]
@cleidigh are you on windows?

**Comment by cleidigh** [comment]
sorry made the trek downstairs and was out of pocket
yes - Windows 7 under an OS X VM (which I can't really use - no decent voice recognition)
Latest Chrome debugger

Unfortunately I can only directly contribute and test Windows VSC

**Comment by misoguy** [comment]
@cleidigh Great job! I have been following this PR but didn't really get the time to review it deeply till now :( It looks great 👍

**Comment by cleidigh** [comment]
@misoguy 
Thanks a bunch!
I think your navigation and the view size keyboard features will be a fine pair coming up together
in the next ;-)

right now I'm looking at what next to attack, how about you?
one of the things would be some more keyboard accessibility for
debugging sessions

**Pull request merged**